### ISSUE TITLE
Some questions about the TranAD paper and code

### ISSUE THREAD

**Issue by chenzijin2018** [imperial-qore/TranAD#70]
Hi:
My name is Zijin Chen. Our team read your paper “TranAD: Deep Transformer Networks for Anomaly Detection in Multivariate Time Series Data” and think the method your team proposed is an great method to detect time series anomaly. And thank you very much for sharing the code. Here we have some questions about the code. Hope you can clarify them. Your answers are much appreciated. Thank you very much.

1.	In class TranAD function encode
src = src * math.sqrt(self.n_feats). We know in the attention. We will divide math.sqrt(self.n_feats). Here we multiply math.sqrt(self.n_feats). Is this trying to enhance the features input signal? Why is math.sqrt(self.n_feats). Have you tried to time another scalar?

2.	In the paper, “We modify the self-attention in the window encoder to mask the data at subsequent positions.” But in class TranAD. We didn’t find where the mask is added. Could you please clarify it?

3.	Our source input is window dimensions(10)*batch size*feature dimensions. Then we concatenate with focus score, now we have input window dimensions(10)*batch size*(2*feature dimensions). Our target is last timestamp(the 10th) record in the window, 1*batch size*feature dimensions. Then we repeat our target to get 1*batch size*(2*feature dimensions). Is this trying to enhance the target features’ signal? The reason I am asking is now. source is features + focus score, while target is features + features. Will this cause any confusion when we try to decode target(features + features) using the encode representation key and value training from features + focus score?

4.	In the paper, we have this Adversarial Training. But in the class TranAD code. I didn’t find where we define  . In the code, we have x1 = self.fcn(self.transformer_decoder1(*self.encode(src, c, tgt))) and x2 = self.fcn(self.transformer_decoder2(*self.encode(src, c, tgt))). My understanding x1 is  , x2 is   in the “Figure 1: The TranAD Model”. Our loss function is “(1 / n) * l(z[0], elem) + (1 - 1/n) * l(z[1], elem)” which is L1 in the paper. Here z[0]=x1,  z[1]=x2.   
My question is where L2 is defined. When we train the model, we can only have one loss function, so we need to have some way to combine L1 and L2, right? Otherwise, we will train two different models, but these two models share the same weights for the same model structure. How do you do that? Do you just sum L1 and L2 or do some weighted average of L1 and L2. If weighted average, how do you determine the weight?

5.	Have you tried Pre-LN Transformer layer instead of Post-LN Transformer layer?

**Comment by shreshthtuli** [comment]
Thanks for the questions.
1. This is for scaling the activations and make training more robust.
2. See dlutils.py.
3. Yes, this is to ensure focus score informed inference. We observed that this improves performance and does not confuse the model.
4. We used the sum of L1 and L2. 
5. We tried both, but they were not as performant for this problem in the datasets we used.
Hope this helps.

**Comment by MKrbm** [comment]
I also have the same questions as 3 and 4. Especially they looked like there are inconsistencies between Fig1 and here.
